File: packages/neptune_flutter_ui/lib/src/brandprint/codec.dart
// © 2026 Neptune.Fintech (neptune.ly) · Neptune Odyssey Community License v1.0
//
// Brandprint codec (Dart port). Faithful, byte-identical port of the TypeScript
// reference (packages/neptune_tokens/src/brandprint/codec.ts). 28-byte fixed
// layout -> base64url, version "NO1-", checksummed. Golden-tested against the
// four reference brands. See docs/11-config-hash.md for the wire format.

import 'dart:convert';
import 'dart:typed_data';

/// Append-only font registry. Indices ARE the wire format — never reorder.
const List<String> kFonts = [
  'Hanken Grotesk',
  'Bricolage Grotesque',
  'Space Grotesk',
  'Sora',
  'IBM Plex Sans Arabic',
  'Reem Kufi',
  'Tajawal',
  'Readex Pro',
  'Noto Kufi Arabic',
];

/// Append-only login-shell registry. The last two (2.24.0) are the shells a
/// bank picks when the pre-login moment is the lockup on a plain ground:
/// `paper-lockup` (light surface, the lockup centred and small, no watermark,
/// no motif) and `lockup-rule` (white ground, the lockup, one hairline rule
/// under it echoing the wordmark's baseline).
const List<String> kLoginShells = [
  'depth-emblem',
  'arcade-arches',
  'light-grid-spark',
  'shield-guilloche',
  'paper-lockup',
  'lockup-rule',
  // 2.29.0. The lockup on a luminous depth field rather than on a plane:
  // the brand gradient travelling under a bloom, framed by one arc. For a
  // bank whose identity is light arriving, which neither a flat canvas nor
  // paper can state.
  'tide-depth',
];

/// Append-only dashboard-hero registry. The last two (2.24.0):
/// `statement-ledger` (one tabular balance statement, then compact account
/// rows - no carousel) and `chevron-summary` (the total across the top, each
/// account row carrying a movement chevron in the brand's accent).
const List<String> kDashboardHeroes = [
  'balance-cards',
  'warm-balance-cards',
  'wallet-hero',
  'restrained-balance',
  'statement-ledger',
  'chevron-summary',
  // 2.29.0. A carousel again, but in the depth register: cards that travel
  // from a deep tone to a bright one with the brand's own wave planes struck
  // across them, sized so the next card peeks in at the reading edge. The
  // peek is the lever's point - `balance-cards` fills the viewport, so a
  // customer with four accounts has to be TOLD there are four.
  'tide-carousel',
];

/// Append-only content-tone registry.
const List<String> kContentTones = [
  'clear-calm',
  'warm-hospitable',
  'light-instant',
  'formal-authoritative',
];

/// Append-only glass-tint registry.
const List<String> kGlassTints = [
  'oceanic',
  'warm-amber',
  'violet-luminous',
  'navy-steel',
];

/// Append-only motion registry.
const List<String> kMotions = [
  'smooth-fluid',
  'calm-graceful',
  'light-quick-crisp',
  'stable-minimal-authoritative',
];

/// Append-only motif registry - byte 26, which was reserved (always `0`)
/// until 2.24.0. Index 0 is `auto`: derive the motif from `glassTint` exactly
/// as before the byte was claimed, so every brandprint already in the wild
/// decodes to the identical theme. The named entries decouple the motif from
/// the glass recipe; `none` is a brand with no pattern at all.
const List<String> kMotifs = [
  'auto',
  'sonar-rings',
  'coastal-arcs',
  'grid-spark',
  'guilloche',
  'none',
];

/// Append-only navigation-shell registry - flags bits 4-5 (2.28.0). The bar a
/// bank's signed-in app stands on, which until now every bank inherited from
/// Andalus:
///
/// * `raised-dock` - the floating glass pill, the active item lifted into a
///   filled circle. Index 0, so every brandprint already in the wild decodes
///   to exactly the dock it has today.
/// * `register-bar` - a flat, full-width bar on one `outlineVariant`
///   hairline. No pill, no float, no fill: the active item is marked by
///   weight and the brand colour, nothing else.
/// * `rule-bar` - a full-width bar under a rule, the active item marked by a
///   segment of that rule in the brand's accent. Structure drawn in lines.
///
/// FOUR ENTRIES MAX: this registry is two bits on the wire (the 28-byte
/// layout has no spare byte left), so a fifth shell needs a format bump, not
/// another list entry.
const List<String> kNavShells = [
  'raised-dock',
  'register-bar',
  'rule-bar',
  // 2.29.0, AND THE LAST ENTRY THIS REGISTRY CAN TAKE - two bits, four values,
  // full. A fifth navigation shell needs a format bump, not another line here.
  //
  // `centre-dock` is the raised pill with a hole in the middle of it: the
  // bank's primary verb has left the quick-action row at the top of the page
  // and become a circle in the bar, reachable with one thumb from any tab.
  // The dock still does not own that button (see [NeptuneDock.centerGap]) -
  // it reserves the space and the host stacks its own action over it, because
  // which verb is primary is the host's decision and the route behind it is
  // the host's too.
  'centre-dock',
];

/// Append-only quick-action-row registry - flags bits 6-7 (2.28.0). The
/// treatment behind a home quick action:
///
/// * `filled-circles` - the tonal `secondaryContainer` circle behind every
///   glyph. Index 0: what every existing brandprint already draws.
/// * `register-rows` - no chip at all. One strip ruled top and bottom, the
///   actions divided by hairlines, like a column header in a ledger.
/// * `rule-grid` - each action in its own hairline cell, and the FIRST action
///   - the one that moves the customer forward - carries the brand accent.
///
/// FOUR ENTRIES MAX, for the same reason as [kNavShells].
const List<String> kActionRows = [
  'filled-circles',
  'register-rows',
  'rule-grid',
];

/// An OKLCH seed colour (perceptual lightness, chroma, hue degrees).
class Seed {
  final double l;
  final double c;
  final int h;
  const Seed({required this.l, required this.c, required this.h});

  @override
  bool operator ==(Object other) =>
      other is Seed && other.l == l && other.c == c && other.h == h;
  @override
  int get hashCode => Object.hash(l, c, h);
}

/// The six corner radii (px), xs..xxl.
class Corners {
  final int xs, sm, md, lg, xl, xxl;
  const Corners({
    required this.xs,
    required this.sm,
    required this.md,
    required this.lg,
    required this.xl,
    required this.xxl,
  });

  @override
  bool operator ==(Object other) =>
      other is Corners &&
      other.xs == xs &&
      other.sm == sm &&
      other.md == md &&
      other.lg == lg &&
      other.xl == xl &&
      other.xxl == xxl;
  @override
  int get hashCode => Object.hash(xs, sm, md, lg, xl, xxl);
}

/// A complete brandprint input config — the inputs a theme is generated from.
/// Mirrors `BrandprintConfig` in codec.ts field-for-field.
class BrandprintConfig {
  final int version;
  final Seed primary;
  final Seed tertiary;
  final Corners corners;
  final int displayWeight;

  /// Tracking in em, e.g. -0.02.
  final double displayTracking;
  final String fontDisplay;
  final String fontText;
  final String fontNum;
  final String loginShell;
  final String dashboardHero;
  final String contentTone;
  final String glassTint;
  final String motion;

  /// One of [kMotifs]. `auto` (the default, and what every pre-2.24.0 string
  /// decodes to) keeps the motif keyed on [glassTint].
  final String motif;
  final bool defaultDark;
  final bool defaultRtl;

  /// Flags bit 2 (2.24.0). When true the [tertiary] seed is the brand's
  /// ACCENT - spent on direction and confirmation only (`NptColors.accent`:
  /// the forward CTA, the active step, an upward movement) - and it feeds no
  /// Material role: the `tertiary*` roles and the card gradient are generated
  /// from [primary] instead, so the accent cannot leak into chrome. False,
  /// which every pre-2.24.0 string decodes to, keeps tertiary as a second
  /// chrome colour and makes the accent equal to primary.
  final bool accentOnTertiary;

  /// Flags bit 3 (2.25.0). The WHITE, STRUCTURAL register: the page ground
  /// and the app bar are `surfaceContainerLowest` (pure white in a light
  /// scheme) instead of `surface` (the tinted tone 98), and a text field is
  /// white inside its ring instead of filled with `surfaceContainerHighest`,
  /// so structure is drawn in lines rather than in grey slabs. Dark mode is
  /// untouched - there is no white to be. False, which every pre-2.25.0
  /// string decodes to, keeps the tinted ground and the filled fields.
  final bool whiteGround;

  /// One of [kNavShells], flags bits 4-5 (2.28.0). The signed-in bar. Index 0
  /// (`raised-dock`) is what every pre-2.28.0 string decodes to.
  final String navShell;

  /// One of [kActionRows], flags bits 6-7 (2.28.0). The home quick-action
  /// treatment. Index 0 (`filled-circles`) is what every pre-2.28.0 string
  /// decodes to.
  final String actionRow;

  const BrandprintConfig({
    this.version = 1,
    required this.primary,
    required this.tertiary,
    required this.corners,
    required this.displayWeight,
    required this.displayTracking,
    required this.fontDisplay,
    required this.fontText,
    required this.fontNum,
    required this.loginShell,
    required this.dashboardHero,
    required this.contentTone,
    required this.glassTint,
    required this.motion,
    this.motif = 'auto',
    this.defaultDark = false,
    this.defaultRtl = false,
    this.accentOnTertiary = false,
    this.whiteGround = false,
    this.navShell = 'raised-dock',
    this.actionRow = 'filled-circles',
  });

  @override
  bool operator ==(Object other) =>
      other is BrandprintConfig &&
      other.version == version &&
      other.primary == primary &&
      other.tertiary == tertiary &&
      other.corners == corners &&
      other.displayWeight == displayWeight &&
      other.displayTracking == displayTracking &&
      other.fontDisplay == fontDisplay &&
      other.fontText == fontText &&
      other.fontNum == fontNum &&
      other.loginShell == loginShell &&
      other.dashboardHero == dashboardHero &&
      other.contentTone == contentTone &&
      other.glassTint == glassTint &&
      other.motion == motion &&
      other.motif == motif &&
      other.defaultDark == defaultDark &&
      other.defaultRtl == defaultRtl &&
      other.accentOnTertiary == accentOnTertiary &&
      other.whiteGround == whiteGround &&
      other.navShell == navShell &&
      other.actionRow == actionRow;

  @override
  int get hashCode => Object.hashAll([
        version,
        primary,
        tertiary,
        corners,
        displayWeight,
        displayTracking,
        fontDisplay,
        fontText,
        fontNum,
        loginShell,
        dashboardHero,
        contentTone,
        glassTint,
        motion,
        motif,
        defaultDark,
        defaultRtl,
        accentOnTertiary,
        whiteGround,
        navShell,
        actionRow,
      ]);
}

/// Encode/decode the portable `NO1-…` brandprint string.
class Brandprint {
  Brandprint._();

  static const int version = 1;
  static const String _prefix = 'NO1-';
  static const int _payloadBytes = 28;

  static int _ix(List<String> arr, String v) {
    final i = arr.indexOf(v);
    return i < 0 ? 0 : i;
  }

  static String _toBase64Url(Uint8List bytes) =>
      base64Url.encode(bytes).replaceAll('=', '');

  static Uint8List _fromBase64Url(String s) {
    var t = s;
    while (t.length % 4 != 0) {
      t += '=';
    }
    return base64Url.decode(t);
  }

  /// Encode a config to its `NO1-…` brandprint string.
  static String encode(BrandprintConfig cfg) {
    final buf = Uint8List(_payloadBytes);
    final dv = ByteData.view(buf.buffer);
    var o = 0;
    buf[o++] = version;
    buf[o++] = (cfg.primary.l * 255).round();
    buf[o++] = (cfg.primary.c * 1000).round().clamp(0, 255);
    dv.setUint16(o, cfg.primary.h, Endian.big);
    o += 2;
    buf[o++] = (cfg.tertiary.l * 255).round();
    buf[o++] = (cfg.tertiary.c * 1000).round().clamp(0, 255);
    dv.setUint16(o, cfg.tertiary.h, Endian.big);
    o += 2;
    final c = cfg.corners;
    for (final v in [c.xs, c.sm, c.md, c.lg, c.xl, c.xxl]) {
      buf[o++] = v.clamp(0, 255);
    }
    buf[o++] = (cfg.displayWeight / 100).round();
    dv.setInt8(o, (cfg.displayTracking * 1000).round());
    o += 1;
    buf[o++] = _ix(kFonts, cfg.fontDisplay);
    buf[o++] = _ix(kFonts, cfg.fontText);
    buf[o++] = _ix(kFonts, cfg.fontNum);
    buf[o++] = _ix(kLoginShells, cfg.loginShell);
    buf[o++] = _ix(kDashboardHeroes, cfg.dashboardHero);
    buf[o++] = _ix(kContentTones, cfg.contentTone);
    buf[o++] = _ix(kGlassTints, cfg.glassTint);
    buf[o++] = _ix(kMotions, cfg.motion);
    var f = 0;
    if (cfg.defaultDark) f |= 1;
    if (cfg.defaultRtl) f |= 2;
    if (cfg.accentOnTertiary) f |= 4;
    if (cfg.whiteGround) f |= 8;
    // The high nibble carries the two composition shells, two bits each. The
    // 28-byte layout is full, and both registries default to index 0, so
    // every brandprint issued before 2.28.0 encodes and decodes unchanged.
    f |= (_ix(kNavShells, cfg.navShell) & 3) << 4;
    f |= (_ix(kActionRows, cfg.actionRow) & 3) << 6;
    buf[o++] = f;
    buf[o++] = _ix(kMotifs, cfg.motif);
    var sum = 0;
    for (var i = 0; i < o; i++) {
      sum = (sum + buf[i]) & 255;
    }
    buf[o++] = sum; // checksum
    return _prefix + _toBase64Url(buf);
  }

  /// Decode a `NO1-…` brandprint string.
  /// Throws [FormatException] on bad prefix/length/checksum/version.
  static BrandprintConfig decode(String str) {
    if (!str.startsWith(_prefix)) {
      throw const FormatException('bad prefix');
    }
    final buf = _fromBase64Url(str.substring(4));
    if (buf.length != _payloadBytes) {
      throw const FormatException('bad length');
    }
    final dv = ByteData.view(buf.buffer, buf.offsetInBytes, buf.lengthInBytes);
    var sum = 0;
    for (var i = 0; i < 27; i++) {
      sum = (sum + buf[i]) & 255;
    }
    if (sum != buf[27]) {
      throw const FormatException('checksum mismatch');
    }
    var o = 0;
    final ver = buf[o++];
    if (ver != version) {
      throw FormatException('version $ver unsupported');
    }
    final primary = Seed(
      l: buf[o++] / 255,
      c: buf[o++] / 1000,
      h: dv.getUint16((o += 2) - 2, Endian.big),
    );
    final tertiary = Seed(
      l: buf[o++] / 255,
      c: buf[o++] / 1000,
      h: dv.getUint16((o += 2) - 2, Endian.big),
    );
    final corners = Corners(
      xs: buf[o++],
      sm: buf[o++],
      md: buf[o++],
      lg: buf[o++],
      xl: buf[o++],
      xxl: buf[o++],
    );
    final displayWeight = buf[o++] * 100;
    final displayTracking = dv.getInt8(o) / 1000;
    o += 1;
    final fontDisplay = kFonts[buf[o++]];
    final fontText = kFonts[buf[o++]];
    final fontNum = kFonts[buf[o++]];
    final loginShell = kLoginShells[buf[o++]];
    final dashboardHero = kDashboardHeroes[buf[o++]];
    final contentTone = kContentTones[buf[o++]];
    final glassTint = kGlassTints[buf[o++]];
    final motion = kMotions[buf[o++]];
    final f = buf[o++];
    final motif = kMotifs[buf[o++]];
    return BrandprintConfig(
      version: ver,
      primary: primary,
      tertiary: tertiary,
      corners: corners,
      displayWeight: displayWeight,
      displayTracking: displayTracking,
      fontDisplay: fontDisplay,
      fontText: fontText,
      fontNum: fontNum,
      loginShell: loginShell,
      dashboardHero: dashboardHero,
      contentTone: contentTone,
      glassTint: glassTint,
      motion: motion,
      motif: motif,
      defaultDark: (f & 1) != 0,
      defaultRtl: (f & 2) != 0,
      accentOnTertiary: (f & 4) != 0,
      whiteGround: (f & 8) != 0,
      navShell: kNavShells[((f >> 4) & 3).clamp(0, kNavShells.length - 1)],
      actionRow: kActionRows[((f >> 6) & 3).clamp(0, kActionRows.length - 1)],
    );
  }
}

File: packages/neptune_tokens/src/brandprint/codec.ts
// Neptune Odyssey — brandprint codec (TypeScript port) · © 2026 Neptune.Fintech (neptune.ly)
// Faithful, byte-identical port of the Dart codec (neptune_flutter_ui). 28-byte layout
// (version byte 1) or, from 2.28.0, 29 bytes (version byte 2) when the extension byte
// carries something -> base64url, prefix "NO1-", checksummed. Golden-tested.
// See docs/11-config-hash.md for the wire format.

import {
  FONTS,
  LOGIN,
  HERO,
  TONE,
  GLASS,
  MOTION,
  MOTIF,
  type Font,
  type LoginShell,
  type DashboardHero,
  type ContentTone,
  type GlassTint,
  type Motion,
  type Motif,
  NAV_SHELL,
  type NavShell,
  ACTION_ROW,
  type ActionRow,
} from "./registries.js";

export interface Seed {
  L: number;
  C: number;
  H: number;
}

export interface Corners {
  xs: number;
  sm: number;
  md: number;
  lg: number;
  xl: number;
  xxl: number;
}

export interface BrandprintConfig {
  primary: Seed;
  tertiary: Seed;
  corners: Corners;
  displayWeight: number;
  /** em, e.g. -0.02 */
  displayTracking: number;
  fonts: { display: Font; text: Font; num: Font };
  loginShell: LoginShell;
  dashboardHero: DashboardHero;
  contentTone: ContentTone;
  glassTint: GlassTint;
  motion: Motion;
  /** Byte 26 (reserved until 2.24.0). Omitted = "auto": derive from glassTint. */
  motif?: Motif;
  defaultDark: boolean;
  defaultRtl: boolean;
  /**
   * Flags bit 2 (2.24.0). True: the tertiary seed is the brand's ACCENT, spent on
   * direction and confirmation only, and feeds no Material role (tertiary* and the
   * card gradient come from primary). Omitted/false: tertiary is a second chrome
   * colour as before and the accent equals primary.
   */
  accentOnTertiary?: boolean;
  /**
   * Flags bit 3 (2.25.0). True: the white, structural register - the page ground
   * and app bar are surface-container-lowest (pure white in light), and a text
   * field is white inside its ring instead of filled. Omitted/false keeps the
   * tinted ground and the filled fields; dark mode is untouched either way.
   */
  whiteGround?: boolean;
  /**
   * Flags bits 4-5 (2.28.0). The bar the signed-in app stands on. Omitted =
   * "raised-dock", the floating glass dock every brandprint had before the
   * nibble was claimed.
   */
  navShell?: NavShell;
  /**
   * Flags bits 6-7 (2.28.0). The home quick-action treatment. Omitted =
   * "filled-circles", the tonal circle behind every glyph.
   */
  actionRow?: ActionRow;
  /**
   * Byte 27 (the extension byte), bit 0 (2.28.0). THE RULED REGISTER: this brand
   * draws structure in LINES - ruled rectangular buttons at the brand's own `md`
   * corner instead of stadium pills, and hairline-ruled groups instead of
   * tone-filled cards. It does NOT ride the flags byte: bits 4-7 are the two
   * composition registries and byte 26 is the motif, so the 28-byte layout was
   * full. Setting it grows the payload to 29 bytes and the version byte to 2;
   * omitted/false encodes to exactly the 28 bytes it always did.
   */
  ruledRegister?: boolean;
}

export interface DecodedBrandprint extends BrandprintConfig {
  version: number;
  motif: Motif;
  accentOnTertiary: boolean;
  whiteGround: boolean;
  navShell: NavShell;
  actionRow: ActionRow;
  ruledRegister: boolean;
}

/** The version byte of the original 28-byte layout. */
export const VERSION = 1;
/** The version byte of the 29-byte layout, which carries the extension byte. */
export const VERSION_EXTENDED = 2;
const PREFIX = "NO1-";
const PAYLOAD_BYTES = 28;
const PAYLOAD_BYTES_EXTENDED = 29;

const ix = <T>(arr: readonly T[], v: T): number => {
  const i = arr.indexOf(v);
  return i < 0 ? 0 : i;
};

function toBase64Url(bytes: Uint8Array): string {
  let bin = "";
  for (let i = 0; i < bytes.length; i++) bin += String.fromCharCode(bytes[i]!);
  const b64 =
    typeof btoa !== "undefined"
      ? btoa(bin)
      : Buffer.from(bytes).toString("base64");
  return b64.replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function fromBase64Url(s: string): Uint8Array {
  let t = s.replace(/-/g, "+").replace(/_/g, "/");
  while (t.length % 4) t += "=";
  if (typeof atob !== "undefined") {
    const bin = atob(t);
    const out = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
    return out;
  }
  return new Uint8Array(Buffer.from(t, "base64"));
}

/** Encode a config to its `NO1-…` brandprint string. */
export function encode(cfg: BrandprintConfig): string {
  // The extension byte is written only when it would carry something, so a config
  // that predates it produces exactly the 28 bytes it always did.
  let ext = 0;
  if (cfg.ruledRegister) ext |= 1;
  const extended = ext !== 0;
  const buf = new Uint8Array(extended ? PAYLOAD_BYTES_EXTENDED : PAYLOAD_BYTES);
  const dv = new DataView(buf.buffer);
  let o = 0;
  buf[o++] = extended ? VERSION_EXTENDED : VERSION;
  buf[o++] = Math.round(cfg.primary.L * 255);
  buf[o++] = Math.min(255, Math.round(cfg.primary.C * 1000));
  dv.setUint16(o, cfg.primary.H);
  o += 2;
  buf[o++] = Math.round(cfg.tertiary.L * 255);
  buf[o++] = Math.min(255, Math.round(cfg.tertiary.C * 1000));
  dv.setUint16(o, cfg.tertiary.H);
  o += 2;
  for (const k of ["xs", "sm", "md", "lg", "xl", "xxl"] as const) {
    buf[o++] = Math.min(255, cfg.corners[k]);
  }
  buf[o++] = Math.round(cfg.displayWeight / 100);
  dv.setInt8(o, Math.round(cfg.displayTracking * 1000));
  o += 1;
  buf[o++] = ix(FONTS, cfg.fonts.display);
  buf[o++] = ix(FONTS, cfg.fonts.text);
  buf[o++] = ix(FONTS, cfg.fonts.num);
  buf[o++] = ix(LOGIN, cfg.loginShell);
  buf[o++] = ix(HERO, cfg.dashboardHero);
  buf[o++] = ix(TONE, cfg.contentTone);
  buf[o++] = ix(GLASS, cfg.glassTint);
  buf[o++] = ix(MOTION, cfg.motion);
  let f = 0;
  if (cfg.defaultDark) f |= 1;
  if (cfg.defaultRtl) f |= 2;
  if (cfg.accentOnTertiary) f |= 4;
  if (cfg.whiteGround) f |= 8;
  // The high nibble carries the two composition shells, two bits each: the
  // 28-byte layout has no spare byte, and index 0 on both keeps every string
  // issued before 2.28.0 byte-identical.
  f |= (ix(NAV_SHELL, cfg.navShell ?? "raised-dock") & 3) << 4;
  f |= (ix(ACTION_ROW, cfg.actionRow ?? "filled-circles") & 3) << 6;
  buf[o++] = f;
  buf[o++] = ix(MOTIF, cfg.motif ?? "auto"); // motif (byte 26)
  if (extended) buf[o++] = ext; // byte 27, the extension byte
  let sum = 0;
  for (let i = 0; i < o; i++) sum = (sum + buf[i]!) & 255;
  buf[o++] = sum; // checksum
  return PREFIX + toBase64Url(buf);
}

/** Decode a `NO1-…` brandprint string. Throws on bad prefix/length/checksum/version. */
export function decode(str: string): DecodedBrandprint {
  if (!str.startsWith(PREFIX)) throw new Error("bad prefix");
  const buf = fromBase64Url(str.slice(4));
  if (buf.length !== PAYLOAD_BYTES && buf.length !== PAYLOAD_BYTES_EXTENDED) {
    throw new Error("bad length");
  }
  const dv = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
  const last = buf.length - 1; // the checksum is always the final byte
  let sum = 0;
  for (let i = 0; i < last; i++) sum = (sum + buf[i]!) & 255;
  if (sum !== buf[last]) throw new Error("checksum mismatch");
  let o = 0;
  const version = buf[o++]!;
  // The version byte NAMES the length. Accepting a mismatch would let a truncated
  // or padded payload decode as a plausible neighbour.
  const expected =
    version === VERSION ? PAYLOAD_BYTES : version === VERSION_EXTENDED ? PAYLOAD_BYTES_EXTENDED : -1;
  if (expected !== buf.length) {
    throw new Error(`version ${version} unsupported at ${buf.length} bytes`);
  }
  const primary: Seed = { L: buf[o++]! / 255, C: buf[o++]! / 1000, H: dv.getUint16((o += 2, o - 2)) };
  const tertiary: Seed = { L: buf[o++]! / 255, C: buf[o++]! / 1000, H: dv.getUint16((o += 2, o - 2)) };
  const corners = {} as Corners;
  for (const k of ["xs", "sm", "md", "lg", "xl", "xxl"] as const) corners[k] = buf[o++]!;
  const displayWeight = buf[o++]! * 100;
  const displayTracking = dv.getInt8(o) / 1000;
  o += 1;
  const fonts = {
    display: FONTS[buf[o++]!] as Font,
    text: FONTS[buf[o++]!] as Font,
    num: FONTS[buf[o++]!] as Font,
  };
  const loginShell = LOGIN[buf[o++]!] as LoginShell;
  const dashboardHero = HERO[buf[o++]!] as DashboardHero;
  const contentTone = TONE[buf[o++]!] as ContentTone;
  const glassTint = GLASS[buf[o++]!] as GlassTint;
  const motion = MOTION[buf[o++]!] as Motion;
  const f = buf[o++]!;
  const motif = MOTIF[buf[o++]!] as Motif;
  // Absent on a 28-byte payload, which is exactly how every pre-2.28.0 string
  // decodes to the defaults it always had.
  const ext = buf.length === PAYLOAD_BYTES_EXTENDED ? buf[o++]! : 0;
  return {
    version,
    primary,
    tertiary,
    corners,
    displayWeight,
    displayTracking,
    fonts,
    loginShell,
    dashboardHero,
    contentTone,
    glassTint,
    motion,
    defaultDark: !!(f & 1),
    defaultRtl: !!(f & 2),
    motif,
    accentOnTertiary: !!(f & 4),
    whiteGround: !!(f & 8),
    navShell: (NAV_SHELL[(f >> 4) & 3] ?? NAV_SHELL[0]) as NavShell,
    actionRow: (ACTION_ROW[(f >> 6) & 3] ?? ACTION_ROW[0]) as ActionRow,
    ruledRegister: !!(ext & 1),
  };
}
